### PR TITLE
User authentication

### PR DESCRIPTION
**Pull request by stefanvulpe-dev**

### CODE REVIEW

**Comment by stefanvulpe-dev** [comment]
Am testat in postman si e okay, returneaza token-ul sau erori in caz ca username-ul nu e bun / nu-l gaseste in baza de date. Also am adaugat si nodemon in dependente ca ma durea mana la cat de mult restartam serverul manual :)).

**Comment by stefanvulpe-dev** [comment]
Am testat in postman si e okay, imi genereaza token-ul sau o eroare in caz ca user-ul nu exista / parola e incorecta.

File: src/controllers/UsersController.ts [hunk 1]
```diff
@@ -2,16 +2,54 @@ import {
...
+    const result = await new UsersRepository().findByUserName(
+      credentials.username,
+    );
+
+    if (!result.user) {
+      throw new Error(`User ${credentials.username} has not been found`);
+    }
+
+    /* Temporary password checking */
+    if (result.user.password !== credentials.password) {
```
Comment (hututudor) [inline src/controllers/UsersController.ts]: Please use something like `bcrypt` (https://www.npmjs.com/package/bcrypt) to check a password hash. We must never store password in clear text in the DB.

**Review by hututudor — APPROVED** [review]

File: src/controllers/UsersController.ts (hunk 1 above)
Comment (stefanvulpe-dev) [reply src/controllers/UsersController.ts]: Yeah, sure, this was just a sample code to get the things up and running at a low level.

File: src/controllers/UsersController.ts [hunk 2]
```diff
@@ -72,7 +79,25 @@ export const auth = (
...
+    return Response.badRequest({ message: 'Missing auhentication token' });
+  }
+
+  try {
+    const decoded = verify(token, process.env.ACCESS_SECRET_KEY!);
+    /* We should store the payload (user info) somewhere, instead of decoding the token on every request */
+    /* e.g. req.user = JSON.parse(decoded.user); */
+    next();
+  } catch (err) {
+    res.redirect('/login');
```
Comment (hututudor) [inline src/controllers/UsersController.ts]: Don't redirect here ... it doesn't have any meaning in a REST api

**Review by hututudor — APPROVED** [review]

**Comment by stefanvulpe-dev** [comment]
@hututudor I'm still waiting on an answer regarding storing user information in the request object.

File: src/controllers/UsersController.ts [hunk 3]
```diff
@@ -34,7 +79,23 @@ export const auth = (
...
+  const { token } = req.body;
+
+  if (!token) {
+    return Response.badRequest({ message: 'Missing auhentication token' });
+  }
+
+  try {
+    const decoded = verify(token, process.env.ACCESS_SECRET_KEY!);
+    /* We should store the payload (user info) somewhere, instead of decoding the token on every request */
+    /* e.g. req.user = JSON.parse(decoded.user); */
```
Comment (hututudor) [inline src/controllers/UsersController.ts]: My bad, we need to verify the token on every request ... and you are right about storing the user object (maybe only the id) in the request

I think the steps are:
1. store the user/userId in request `req.user = user` (you may encounter some typescript errors here, in that case use this: https://blog.logrocket.com/extend-express-request-object-typescript/)
2. in our own `Request` object, add the user/userId property
3. store the property from the `handleRequest` function
4. you should be able to access the `req.user` from all controller handlers

You may check that this works by implementing a test `profile`-like get, where you return the id from the decoded token

**Review by hututudor — COMMENTED** [review]
We don't need `lastName` and `firstName` at all, an user should only have `username`.

File: src/index.ts [hunk 4]
```diff
@@ -3,6 +3,7 @@ import 'dotenv/config';
 import cors from 'cors';
 
 import { migrate } from './db';
+import {} from './types/express/index';
```
Comment (hututudor) [inline src/index.ts]:
```suggestion
import './types/express/index';
```

**Review by hututudor — APPROVED** [review]

**Pull request merged**